### PR TITLE
Enable writing migration files in Typescript

### PR DESCRIPTION
**Pull request by gburgett**
* Adds a Typescript definition file at the root
* Adds doc on how to run Typescript migration files using ts-node

### CODE REVIEW

**Comment by TimBeyer** [comment]
Is your editor set up to remove trailing whitespace?  
They are meaningful in markdown for linebreaks and I think the layout now looks a bit too dense.

**Comment by gburgett** [comment]
Whoops!  Not only does my editor remove trailing whitespace, it fails to show whitespace changes in the diff.  Fixed that :)

**Comment by gburgett** [comment]
Any desire to merge this in?  Any other comments?

**Comment by TimBeyer** [comment]
Hey Gordon, 
we definitely intend to merge this.
We've been blocked a bit by CI issues that didn't allow us to run tests since we were relying on some beta API features in setting them up and needed to refactor parts.  
We also wanted to ensure that external PRs can actually go through CI the same way as internal ones so that not every contributor is faced by a big red cross and the text "All checks have failed".

All this takes a bit longer than anticipated, but rest assured that your contributions are very much welcome and that we'll proceed with merging once everything is working again.

**Comment by gburgett** [comment]
I wrote this cheat sheet for my co-workers, maybe y'all could make use of it or adapt it for your documentation:  https://www.cheatography.com/gburgett/cheat-sheets/contentful-migration/

**Comment by Khaledgarbaya** [comment]
Hi @gburgett ,
I am really sorry about the delay, can you resolve the conflict in the README and I will manually run the integration tests locally for now.

Best,
Khaled

**Comment by Khaledgarbaya** [comment]
Hey @gburgett we made some restructuring for the README. Can you resolve the conflict so I can merge your PR.

Best,
Khaled

**Comment by TimBeyer** [comment]
@Khaledgarbaya might be worth checking if https://github.com/watermarkchurch/migration-cli/blob/master/index.d.ts contains more up to date typings

File: README.md [hunk 1]
```diff
@@ -481,7 +486,35 @@ migrations()
 
 ### You need to share confidential information or have other questions?
 
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (Khaledgarbaya) [reply README.md]: 🤦‍♂️

**Comment by gburgett** [comment]
Hi guys!  I've been trying to keep our fork up to date but y'all have been moving super fast.  I'll have some time to work on this next week, would love to see it get merged in.

BTW the reason we're maintaining a fork is explained in #66, the need for a history of migrations that have already been run.  This was the response,

>  Roblinde commented on Apr 18
> We've had some further discussions internally and the general feeling is that handling this in a content type is not ideal in a broader product perspective. We fully understand the need for a history of migrations run on a specific space and that this needs to be solved for migrations to be a viable option as a part of a robust CI toolchain. This is something we will continue to work on and strive to find an idiomatic solution for.

Did you guys implement something yet?  Could you shed some more light on your future plans for this?

Thanks!

**Comment by Khaledgarbaya** [comment]
Hi @gburgett,
We are looking into ways on having migration history but nothing planned yet.
Let me know if you need help with this PR.

Best,
Khaled

**Comment by Khaledgarbaya** [comment]
Hey @gburgett,
I will have time for this tomorrow to test and play around and if I need to change some stuff I push them directly here.

Best,
Khaled

**Comment by joshmedeski** [comment]
It would be great to see this merged soon, any way I can help?

**Comment by Khaledgarbaya** [comment]
Hey All, I know I've been saying this a lot 😅 but we'll get try to get this merge in the next two days for real this time. We are terribly sorry about the delay

File: README.md [hunk 2]
```diff
@@ -473,6 +473,36 @@ migrations()
 
 ```
 
+## Writing Migrations in Typescript
+
+You can use Typescript to write your migration files using `ts-node`!  First `npm install --save ts-node typescript`,
+then run your migration with ts-node:
+
+```
+node_modules/.bin/ts-node node_modules/.bin/contentful-migration -s $CONTENTFUL_SPACE_ID -a $CONTENTFUL_MANAGEMENT_TOKEN my_migration.ts
```
Comment (Khaledgarbaya) [inline README.md]: I think `ts-node` is missing here at the beginning of the line, or am I missing something?

**Comment by Khaledgarbaya** [comment]
Ok all tests are working fine, I am merging this now, Thank you y'all for the input and for your patience

**Comment by Khaledgarbaya** [comment]
released on `0.12.2`

**Comment by gburgett** [comment]
@Khaledgarbaya my colleague just discovered an issue, the `index.d.ts` is not included in the `files` declaration inside `package.json`.  You will need to add the following patch:

```patch
diff --git package.json package.json
index f6fe178..4b426e9 100644
--- package.json
+++ package.json
@@ -43,7 +43,8 @@
     "CHANGELOG.md",
     "bin",
     "built",
-    "docs"
+    "docs",
+    "index.d.ts"
   ],
   "nyc": {
     "per-file": true,
```

**Comment by Shy** [comment]
Hey @gburgett, as a way of saying thank you for the work in this PR we'd love to send you some Contentful Socks. Would you be able to shoot me an email (shy [at] contentful.com) so I can get your details and we'll mail those to you?

**Pull request merged**